### PR TITLE
Update plugin embedded docs

### LINKED ISSUE #3872: Describe how GeneratedCode is recognized in embedded docs
This seems obvious that it should be, but I cant find documentation stating it is..

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #3872
Fixes background task to document test-code and .editorconfig

### CODE REVIEW

File: sonar-csharp-plugin/src/main/resources/documentation.md [hunk 1]
```diff
@@ -16,12 +16,26 @@ Discover and update the C#-specific [properties](/analysis/analysis-parameters/)
 
 To analyze tool-generated code (e.g. WCF code generated by `SvcUtil.exe`, protobuf code generated by `protoc`, Swagger client code generated by `NSwag`) for a specific C# project, enable the "Analyze generated code" setting inside **Project > Administration > General Settings > C#**. By default, tool-generated code files are skipped from analysis.
 
+Generated code detection is based on file name, special comments and attributes. Currently recognized values are in [GeneratedCodeRecognizer.cs](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs).
```
Comment (andrei-epure-sonarsource) [inline sonar-csharp-plugin/src/main/resources/documentation.md:19]:
```suggestion
The detection of generated code is based on the file name, special comments, and attributes. The currently recognized values are in [GeneratedCodeRecognizer.cs](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs).
```

Comment (andrei-epure-sonarsource) [reply sonar-csharp-plugin/src/main/resources/documentation.md:19]: (BTW I installed the Grammarly browser extension which helps with hints)

File: sonar-csharp-plugin/src/main/resources/documentation.md [hunk 2]
```diff
@@ -16,12 +16,26 @@ Discover and update the C#-specific [properties](/analysis/analysis-parameters/)
...
+
 ## Scanner compatibility
 
 To analyze C# code, you need to use the [SonarScanner for .NET](/analysis/scan/sonarscanner-for-msbuild/) version 4.x or newer.
 
+## Exclusions
+
+Files excluded from analysis should be set in a project configuration. This exclusion still performs the analysis on the file and filters final results.
```
Comment (andrei-epure-sonarsource) [inline sonar-csharp-plugin/src/main/resources/documentation.md:27]:
```suggestion
The files which are to be excluded from analysis should be set in the project configuration. The files are still going to be analyzed during the MSBuild compilation, the exclusion happening as a filter before sending the results to the server.
```

File: sonar-csharp-plugin/src/main/resources/documentation.md [hunk 3]
```diff
@@ -16,12 +16,26 @@ Discover and update the C#-specific [properties](/analysis/analysis-parameters/)
...
+
+As an alternative, `.editorconfig` file can be used to disable analysis for a specific rule on a file or directory. This can solve performance problems on large files.
```
Comment (andrei-epure-sonarsource) [inline sonar-csharp-plugin/src/main/resources/documentation.md:29]:
```suggestion
As an alternative, an `.editorconfig` file can be used to disable the analysis for a specific rule on a file or directory. This can solve performance problems on large files. 
```

File: sonar-csharp-plugin/src/main/resources/documentation.md [hunk 4]
```diff
@@ -16,12 +16,26 @@ Discover and update the C#-specific [properties](/analysis/analysis-parameters/)
...
+
+```
+[Path/File.cs]
+dotnet_diagnostic.Sxxx.severity = none
```
Comment (andrei-epure-sonarsource) [inline sonar-csharp-plugin/src/main/resources/documentation.md:33]: We have a problem that the class name appears in `reportanalyzer` logs and the ID is hard to get... Maybe we should mention the version of Roslyn where this works, and also how to get the rule ID

File: sonar-csharp-plugin/src/main/resources/documentation.md [hunk 5]
```diff
@@ -16,12 +16,26 @@ Discover and update the C#-specific [properties](/analysis/analysis-parameters/)
...
+```
+
```
Comment (andrei-epure-sonarsource) [inline sonar-csharp-plugin/src/main/resources/documentation.md:35]:
```suggestion

You can read more in the [official documentation](https://docs.microsoft.com/en-us/visualstudio/code-quality/use-roslyn-analyzers?view=vs-2019#manually-configure-rule-severity-in-an-editorconfig-file).
```

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]
I left a few comments.

File: sonar-vbnet-plugin/src/main/resources/documentation.md [hunk 6]
```diff
@@ -16,15 +16,30 @@ Discover and update the VB.NET-specific [properties](/analysis/analysis-paramete
...
+dotnet_diagnostic.Sxxx.severity = none
+```
+
+## Known Limitations
+
+Currently an error will be thrown when an issue is raised on a line of code containing the following pattern `\s+error\s*:` (i.e. one or more spaces, the string 'error', zero or more spaces and a ':' ) . This is a well known problem on the Microsoft side (see [issue](https://github.com/dotnet/roslyn/issues/5724/)). In order to work around this problem, analysis will skip issues reported on any line where the pattern is detected.
+
 ## Related Pages
 * [Importing External Issues](/analysis/external-issues/) (VSTest, NUnit, MSTest, xUnit)
 * [Test Coverage & Execution](/analysis/coverage/) (Visual Studio Code Coverage, dotCover, OpenCover, NCover 3)
```
Comment (andrei-epure-sonarsource) [inline sonar-vbnet-plugin/src/main/resources/documentation.md:42]: how about also adding links to our guide: https://community.sonarsource.com/t/9871 and https://community.sonarsource.com/t/37151

File: sonar-csharp-plugin/src/main/resources/documentation.md (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply sonar-csharp-plugin/src/main/resources/documentation.md:27]: Are you sure about the `which are to be`? That one sounds very complicated/future-conditional-ish.

Comment (pavel-mikula-sonarsource) [reply sonar-csharp-plugin/src/main/resources/documentation.md:27]: `the exclusion happening as a` - isn't there a verb missing? `the exclusion is happening as a`

Comment (pavel-mikula-sonarsource) [reply sonar-csharp-plugin/src/main/resources/documentation.md:27]: The `before sending the results to the server` is providing too much technical details IMO.

File: sonar-csharp-plugin/src/main/resources/documentation.md (hunk 4 above)
Comment (pavel-mikula-sonarsource) [reply sonar-csharp-plugin/src/main/resources/documentation.md:33]: I'd keep that as a community guide about performance troubleshooting. This is analyzer docs and /p:ReportAnalyzer is compiler stuff.

File: sonar-csharp-plugin/src/main/resources/documentation.md (hunk 5 above)
Comment (pavel-mikula-sonarsource) [reply sonar-csharp-plugin/src/main/resources/documentation.md:35]: I intentionally avoided that, as it's VS/dotnet stuff that's not specific to our analyzer. And docs show a very easy way how to disable rules contrary your QP/QG settings.

File: sonar-csharp-plugin/src/main/resources/documentation.md (hunk 2 above)
Comment (andrei-epure-sonarsource) [reply sonar-csharp-plugin/src/main/resources/documentation.md:27]: the idea is that 
> This exclusion still performs the analysis on the file and filters final results.
Is not clear at all, because:
- we say "the exclusion performs the analysis", which isn't the case (the exclusion doesn't perform anything)
- either way, our users don't know when the analysis is done, that it is done during the build etc

AND I agree with you it's too much detail :). So a new version:

> The files which are to be excluded from analysis should be set in the project configuration. The files are still going to be analyzed during the MSBuild compilation. The configuration of exclusions is used to filter the analysis results.

File: sonar-vbnet-plugin/src/main/resources/documentation.md (hunk 6 above)
Comment (pavel-mikula-sonarsource) [reply sonar-vbnet-plugin/src/main/resources/documentation.md:42]: I've added the second one. The first one is not so related to the analyzer itself, but rather about .NET tooling around.

File: sonar-csharp-plugin/src/main/resources/documentation.md [hunk 7]
```diff
@@ -16,12 +16,27 @@ Discover and update the C#-specific [properties](/analysis/analysis-parameters/)
...
 
+The detection of generated code is based on file name, special comments and attributes. The currently recognized values are in [GeneratedCodeRecognizer.cs](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs).
+
 ## Scanner compatibility
 
 To analyze C# code, you need to use the [SonarScanner for .NET](/analysis/scan/sonarscanner-for-msbuild/) version 4.x or newer.
 
+## Exclusions
+
+Files excluded from analysis should be set in the project configuration. The files are still going to be analyzed during the compilation and results will be filtered according to the exclusion settings.
```
Comment (andrei-epure-sonarsource) [inline sonar-csharp-plugin/src/main/resources/documentation.md:27]:
```suggestion
The files which are to be excluded from analysis should be set in the project configuration. Excluded files are still going to be analyzed during the compilation and the results will be filtered according to the exclusion settings.
```

Comment (andrei-epure-sonarsource) [reply sonar-csharp-plugin/src/main/resources/documentation.md:27]: (you can also ask @tom-howlett-sonarsource as he's a native English speaker)

File: sonar-vbnet-plugin/src/main/resources/documentation.md (hunk 6 above)
Comment (andrei-epure-sonarsource) [reply sonar-vbnet-plugin/src/main/resources/documentation.md:42]: ok, anyway the troubleshooting guide has a link to them

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM.

I think it's also good to ask @tom-howlett-sonarsource for a review (both from a native speaker and from a PM point of view).

File: sonar-csharp-plugin/src/main/resources/documentation.md [hunk 8]
```diff
@@ -16,12 +16,27 @@ Discover and update the C#-specific [properties](/analysis/analysis-parameters/)
 
 To analyze tool-generated code (e.g. WCF code generated by `SvcUtil.exe`, protobuf code generated by `protoc`, Swagger client code generated by `NSwag`) for a specific C# project, enable the "Analyze generated code" setting inside **Project > Administration > General Settings > C#**. By default, tool-generated code files are skipped from analysis.
 
+The detection of generated code is based on file name, special comments and attributes. The currently recognized values are in [GeneratedCodeRecognizer.cs](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs).
```
Comment (andrei-epure-sonarsource) [inline sonar-csharp-plugin/src/main/resources/documentation.md:19]:
```suggestion
The detection of generated code is based on file name, special comments, and attributes.
```

Comment (andrei-epure-sonarsource) [inline sonar-csharp-plugin/src/main/resources/documentation.md:19]:
```suggestion
The detection of generated code is based on file name, special comments, and attributes. The currently recognized values are in [GeneratedCodeRecognizer.cs](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs).
```

Comment (andrei-epure-sonarsource) [reply sonar-csharp-plugin/src/main/resources/documentation.md:19]: grammarly says there a comma needed there. @tom-howlett-sonarsource could you confirm/infirm?

Comment (tom-howlett-sonarsource) [reply sonar-csharp-plugin/src/main/resources/documentation.md:19]: I believe Grammarly is proposing an [Oxford Comma](https://www.grammarly.com/blog/what-is-the-oxford-comma-and-why-do-people-care-so-much-about-it/) here. I personally wouldn't use one here, feels unnecessary.

File: sonar-csharp-plugin/src/main/resources/documentation.md (hunk 7 above)
Comment (pavel-mikula-sonarsource) [reply sonar-csharp-plugin/src/main/resources/documentation.md]: Files to be exluded....

Comment (tom-howlett-sonarsource) [reply sonar-csharp-plugin/src/main/resources/documentation.md]: Either read fine. The second starts to feel a bit wordy - "Files **to be** excluded should..." works for me

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM
